feat: add targeting schema, improve def. schema

toddbaert opened the pull request:
This PR:

- adds JSON-schema validation for targeting rules
  - basic JSONLogic rules validated
  - custom rules (`sem_ver`, `fractional`, `starts/ends_with`) are fully validated 
  - descriptions are added for all operations and operands
  - `$evalutators` validated
  - maintained in separate schema
- improves flag-definition schema
  - descriptions added/improved

I've also added a bunch of positive and negative tests.

For a preview, add `"$schema": "https://deploy-preview-1115--polite-licorice-3db33c.netlify.app/schema/v0/flagd-definitions.json"` as a sibling to `"flags"` in your flagd config.

Closes: https://github.com/open-feature/flagd-schemas/issues/115

:warning: KEEP IN MIND WHEN REVIEWING that that `.yaml` files are the source of truth; the CI checks that the json is consistent with them.

![validation](https://github.com/open-feature/flagd-schemas/assets/25272906/d29fc404-0014-47ed-b5ad-60255fc124b5)

beeme1mr commented on json/targeting.json:
```diff
@@ -0,0 +1,553 @@
+    },
+    "associativeRule": {
+      "title": "Mathematically Associative Operation",
+      "description": "Pperation applicable to 2 or more parameters.",
```
```suggestion
      "description": "Operation applicable to 2 or more parameters.",
```

I'm not 100% sure this is what you intended.

beeme1mr commented on json/flagd-definitions.json:
Perhaps we should use this as an opportunity to set the default variants for boolean flags to "true" and "false" instead of "on" and "off". That would make it easier for people to leverage the shorthand targeting rules.

toddbaert replied:
I can do that. For now I'd rather not include the rest of the changes [here](https://github.com/open-feature/flagd-schemas/pull/81/files) though.

toddbaert replied:
done.

beeme1mr commented:
VSCode supports a few additional properties like `markdownDescription` that we may want to consider leveraging.

https://code.visualstudio.com/docs/languages/json#_use-rich-formatting-in-hovers

beeme1mr commented on json/targeting.json:
```diff
@@ -0,0 +1,553 @@
+    "targeting": {
+      "title": "Targeting",
+      "description": "An expression returning a value which is coerced to a string to be used as a targeting key, or null (to fall back to defaultVariant). If targeting returns a value which is not a variant key, it's considered an error.",
+      "anyOf": [
+        {
+          "$comment": "we need this to support empty targeting",
+          "type": "object",
+          "additionalProperties": false,
+          "properties": {}
+        },
+        {
+          "$ref": "#/$defs/anyRule"
+        }
```
Unfortunately, it looks like this messes with VSCodes intellisense. Valid operators don't autocomplete until the empty targeting support section is commented out.

toddbaert replied:
I think this is more about how vscode in particular does auto-completion and not about JSON schema. It seems that since `{}` is valid there's not much I can do (though I did try a few things to make vscode behave as we'd want). Frankly I'm not sure even if we got it working that it would stay working...

I think we can either remove the `{}` support (and optionally remove it from flagd also?) or accept that vscode auto-complete won't work for the top-level rule (it works for the rest since `{}` is only acceptable at the root).

If we just remove the `{}` from the schema, it might be acceptable... whether or not we actually remove the support in flagd. I'm not really sure there's much value in supporting it officially. What are your thoghts?

toddbaert replied:
@beeme1mr thinking about it more, I believe we should just not consider empty targeting valid from the schema perspective. We can keep the actual flags implementations as they are. Empty targeting just returns the defaultValue anyway.

beeme1mr replied:
I was also able to get it working if I change "anyRule" to use "anyOf" instead of "oneOf" but I'm not sure what impact that may have.

toddbaert replied:
I'll set if that has any negative impacts.

toddbaert replied:
Actually that seems to work without issue... empty targeting `{}` is still valid, and vscode seems to auto-complete, and all the validation continues to work otherwise. I don't really understand why vscode can't auto-complete with "oneOf" as any rule, but I think there's no downside to this so :shrug:

beeme1mr commented on json/targeting.json:
```diff
@@ -0,0 +1,553 @@
+          "anyOf": [
+            {
+              "type": "string",
+              "description": "flagd automatically injects \"$lagd.timestamp\" (unix epoch) and \"$lagd.flagKey\" (the key of the flag in evaluation) into the context.",
```
```suggestion
              "description": "flagd automatically injects \"$flagd.timestamp\" (unix epoch) and \"$flagd.flagKey\" (the key of the flag in evaluation) into the context.",
```

Kavindu-Dodan commented on json/flagd-definitions.json:
```diff
@@ -1,10 +1,13 @@
 {
+  "$id": "https://flagd.dev/schema/flagd-definitions.json",
```
is this still valid?

Kavindu-Dodan commented on json/flagd-definitions.yaml:
```diff
@@ -1,68 +1,85 @@
+$id: "https://flagd.dev/schema/flagd-definitions.json"
```
is this still valid?

toddbaert replied:
Yes, that is the ID of the schema - it's basically a name. It's ideal if it's also the location of the schema, and the extension doesn't really matter.

toddbaert commented:
> VSCode supports a few additional properties like `markdownDescription` that we may want to consider leveraging.
> 
> https://code.visualstudio.com/docs/languages/json#_use-rich-formatting-in-hovers

Seems like "defaultSnippets" also is an interesting non-standard addition it supports.

beeme1mr commented:
> Seems like "defaultSnippets" also is an interesting non-standard addition it supports.

Wow, that's an interesting feature!

toddbaert commented:
@beeme1mr @Kavindu-Dodan I've added a bunch of tests. There was already some really nice testing implemented that simply iterates over a directory of definitions expected to fail and definitions expected to pass, so I just added a bunch of each!

@beeme1mr I've also changed the defaults for boolean, and used `anyOf` as you suggested (see [here](https://github.com/open-feature/flagd-schemas/pull/120#discussion_r1443210509) for more info). I didn't add any of the vscode-specific features. I don't think it's a bad idea but I'd prefer to do it in a different PR since it's non-breaking and this is already big.

toddbaert commented on json/flagd_definitions.go:
```diff
@@ -4,3 +4,6 @@ import _ "embed"
 var FlagdDefinitions string
+
+//go:embed targeting.json
+var Targeting string
```
Now we export a const for the `Targeting` schema as well.

Kavindu-Dodan replied:
Nice. We should follow up and add validations in flagd for flag parsing [^1]

[^1]: https://github.com/open-feature/flagd/blob/main/core/pkg/evaluator/json.go#L404-L406

toddbaert commented on json/flagd_definitions_test.go:
```diff
@@ -2,19 +2,29 @@ package flagd_definitions_test
 
 func init() {
-	schemaLoader = gojsonschema.NewStringLoader(flagd_definitions.FlagdDefinitions)
+	schemaLoader = gojsonschema.NewSchemaLoader()
+	schemaLoader.AddSchemas(gojsonschema.NewStringLoader(flagd_definitions.Targeting))
+	var err error
+	s, err = schemaLoader.Compile(gojsonschema.NewStringLoader(flagd_definitions.FlagdDefinitions))
+	if err != nil {
+		s := fmt.Errorf("err: %v", err)
+		log.Fatal(s)
+	}
```
@Kavindu-Dodan you asked about this... this is how we can directly import schemas in Go (notice I import the referenced (`$ref`) schema first, and then the root schema. This means no actual http requests will be made, the schemas are cached and referenced by their `$id`. In most prod systems, this is how it must work - few servers or libraries will actually make HTTP requests for schemas, that's typically more of a development feature (like for IDEs).

See [this link](https://json-schema.org/understanding-json-schema/structuring#schema-identification):

> Even though schemas are identified by URIs, those identifiers are not necessarily network-addressable. They are just identifiers. Generally, implementations don't make HTTP requests (https://) or read from the file system (file://) to fetch schemas. Instead, they provide a way to load schemas into an internal schema database. When a schema is referenced by it's URI identifier, the schema is retrieved from the internal schema database.

toddbaert replied:
Note I improved some var-names here since I made this comment.

beeme1mr replied:
I didn't even consider this but it's definitely good to know.

Kavindu-Dodan replied:
Cool, thank you for the link and the comment. I thought this to be a valid URL hence my doubt :)

toddbaert commented on json/test/negative/invalid-flagd-props.json:
```diff
@@ -0,0 +1,23 @@
+      "targeting": {
+        "if": [
+          {
+            "==": [{ "var": "$flagd.myprop" }, { "var": "someprop" }]
```
```suggestion
            "==": [{ "var": "$flagd.invalidProp" }, { "var": "someprop" }]
```

Kavindu-Dodan commented on json/targeting.yaml:
```diff
@@ -0,0 +1,411 @@
+    properties:
+      starts_with:
+        title: Starts-With Operation
+        description: Attribute starts with the specified value.
```
Should we mention that the attribute should be a string? 

```suggestion
        description: The string attribute starts with the specified string value.
```

Kavindu-Dodan commented on json/targeting.yaml:
```diff
@@ -0,0 +1,411 @@
+    - $ref: "#/$defs/anyRule"
+  stringCompareArgs:
+    type: array
+    minItems: 2
```
we can also add `"maxItems": 2` as we define the spec of str_comp arguments in our spec [1^]


[^1]: https://github.com/open-feature/flagd/blob/main/docs/reference/custom-operations/string-comparison-operation.md

toddbaert replied:
good call, added.

Kavindu-Dodan reviewed (approved):
Awesome work 👏 

Added few suggestions but I am approving this anyway.

beeme1mr commented on json/targeting.yaml:
```diff
@@ -0,0 +1,412 @@
+          form accepting more than 3 arguments (if/else) is not supported in flagd;
+          use nesting instead.'
+        type: array
+        minItems: 3
```
```suggestion
        minItems: 2
```

The else block is optional and works fine if omitted.
